jmetal-algorithm/src/main/java/org/uma/jmetal/algorithm/multiobjective/moead/MOEADD.java

```diff
@@ -232,7 +221,7 @@ public void updateArchive(S indiv) {
 
     // find the location of 'indiv'
     setLocation(indiv, idealPoint, nadirPoint);
-    int location = regionSolution.getOrDefault(indiv, 0);
+        int location = (int) indiv.getAttribute("region");
```

Comment:
Same thing everywhere, so I won't repeat it anymore. Please check all your indentation.

Meanwhile, this line is among the relevant changes of the pull request: you get the attributes from the individual itself rather than some global containers. The actual issue with maps is that they store the information *ad vitam eternam*. With your change, the individual stores the information by itself so it disappears with it.

The first question is whether or not we want to store the information indefinitely. If we assume that the same solutions might be generated really often, then it is interesting because we save computation time by not regenerating the same values. But here it seems that it does not happen that often. So you were right by choosing a way which makes the information disappear with the individual. However, you did not need all these changes for that: just replacing `HashMap` by `WeakHashMap` allows to let the garbage collector get rid of individuals not used anymore. So if I am correct, just replacing the three maps by weak maps should be enough to fix the problem.

I did not try it, but I would recommend to try with this small change first to see if it fixes the issue. If it does not, then there is something else going wrong, and this pull request might be worth to merge (once the indentation and generics have been fixed).